Rewrite WindowStateStorage for improved async power

UniversalSuperBox opened the pull request:
6931e29 changed WindowStateStorage from asynchronous writes to synchronous due to changes in QSql that denied reusing connections in different threads. This commit splits out all QSql functionality from the WindowStateStorage object into a separate AsyncQuery object and moves the AsyncQuery object into another thread. This allows performing write operations in another thread, preventing main thread freezes from slow IO.

If database initialization fails, the object simply returns the default values provided and black-holes writes.

Also adds tests to ensure the database name is being used sanely and that SQL injection is guarded against.

Co-Authored-By: Rodney Dawes <dobey@ubports.com>

UniversalSuperBox committed:
Rewrite WindowStateStorage for improved async power

6931e29 changed WindowStateStorage from asynchronous writes to synchronous due to changes in QSql that denied reusing connections in different threads. This commit splits out all QSql functionality from the WindowStateStorage object into a separate AsyncQuery object and moves the AsyncQuery object into another thread. This allows performing write operations in another thread, preventing main thread freezes from slow IO.

If database initialization fails, the object simply returns the default values provided and black-holes writes.

Also adds tests to ensure the database name is being used sanely and that SQL injection is guarded against.

Co-Authored-By: Rodney Dawes <dobey@ubports.com>

UniversalSuperBox commented on plugins/Utils/windowstatestorage.cpp:
```diff
@@ -16,167 +17,319 @@
+    }
+    // This could be a simple connection like all the other save* methods
+    // on AsyncQuery, but WindowStateStorage::WindowState can't be used
+    // on it directly.
```
I think there should be a way to use WindowStateStorage::WindowState as an argument to AsyncQuery::saveState, but I haven't been able to find it yet. It compiles if I flip the order of AsyncQuery and WindowStateStorage in the header file and forward-declare AsyncQuery, but the connect call fails saying that WindowStateStorage::WindowState is not a registered metatype.

UniversalSuperBox replied:
I got it, but I'm not sure if I did it the best way possible. Where should I put the qRegisterMetaType call? https://github.com/ubports/unity8/pull/356/commits/02ca0715eae67635a597bd1ced218107a5b21e86

UniversalSuperBox commented on plugins/Utils/windowstatestorage.cpp:
```diff
@@ -16,167 +17,319 @@
-    return string.remove(QLatin1Char('\"'))
-                 .remove(QLatin1Char('\''))
-                 .remove(QLatin1Char('\\'));
+AsyncQuery::~AsyncQuery()
```
Should AsyncQuery be in its own \_p header file? How do I do that correctly?

mariogrip replied:
Imo we should just place it in on top of the cpp file, it is not shared anywhere/

UniversalSuperBox replied:
I have moved AsyncQuery out of the header file. It is now declared in the .cpp.

UniversalSuperBox commented on plugins/Utils/windowstatestorage.cpp:
```diff
@@ -16,167 +17,319 @@
+        return defaultValue;
     }
+    QRect geometry;
+    QMetaObject::invokeMethod(m_asyncQuery, "getGeometry", Qt::BlockingQueuedConnection,
```
I have a feeling that a read encountering a write in progress would freeze the UI due to the database being locked. Should we time out the queries to prevent that?

UniversalSuperBox replied:
I added a timeout in https://github.com/ubports/unity8/pull/356/commits/209b85132073915f87ba111155056e05d751ec89

mardy replied:
Wait... AFAIK, in SQLite a reader is never blocked; the only operation that can be blocked is the write one.

UniversalSuperBox replied:
In section 5.0 of [File Locking and Concurrency in SQLite Version 3](https://sqlite.org/lockingv3.html), it is specified that there is a short period of time during a write that readers are blocked. So it's totally possible, and it was cheap enough to add the timeout.

mariogrip commented on plugins/Utils/windowstatestorage.cpp:
```diff
@@ -16,167 +17,319 @@
-                 .remove(QLatin1Char('\\'));
+AsyncQuery::~AsyncQuery()
+{
+    QSqlDatabase::removeDatabase(m_connectionName);
```
Is this always safe to run? even if there is no database to remove?

UniversalSuperBox replied:
Yes.

mariogrip commented on plugins/Utils/windowstatestorage.cpp:
```diff
@@ -16,167 +17,319 @@
+    m_thread.start();
+    bool queryInitSuccessful;
+    QMetaObject::invokeMethod(m_asyncQuery, "initdb",
+                              Qt::BlockingQueuedConnection,
```
Does this need to be blocking? And if it some of the funcs get called before its inited, maybe use QWaitCondition or something in the funcs to make sure its inited,

UniversalSuperBox replied:
Well, I suppose it would make Lomiri's start a little faster under conditions of high IO load... but the first application that the user launched would still have to wait on the database to initialize. In other words, you're just moving the delay to first app launch instead of Lomiri launch. Is that worth it?

peat-psuwit replied:
Hmm, it can be some time between this class is init and the user starts an app, so it could be an overall win. I'm not sure about this though, and the complexity factor might make this idea not worth it.

mardy replied:
A Queued connection is enough, really:

> It needs to be synchronous, else it is possible to save a value before the connection is opened. We don't really want that.

This cannot happen, because the thread is running an event loop, and will process all the queued requests in the same order they were sent: so it will always process the `initdb()` method first.

mariogrip commented on plugins/Utils/windowstatestorage.h:
```diff
@@ -47,25 +84,25 @@ class WindowStateStorage: public QObject
-    QSqlDatabase m_db;
+private:
+    QThread m_thread;
+    AsyncQuery *m_asyncQuery;
```
Would like to see a std::shared_ptr here, else you need to make sure it gets destroyed after the class is deconstructed. but we should avoid using raw ptrs. (even if we use .get())

UniversalSuperBox replied:
See https://github.com/ubports/unity8/pull/356/files#diff-95c2b40a9d9f8f86e9324b7ded598588aebd8ef27d61c36021d3e538e492e0c9R232, which will ensure the object is deleted when the thread stops.

mardy replied:
Yes, here using deleteLater is safer, since the object is in another thread

mariogrip reviewed (dismissed):
Some small things

UniversalSuperBox committed:
Move AsyncQuery entirely into cpp file

UniversalSuperBox committed:
Add a one-second timeout to prevent deadlocks

I'm assuming that WindowStateStorage information is _really_ not that
important with this, though one second is an eternity to be waiting on
IO. The busy timeout will ensure that Lomiri only freezes for a read
for 1 second if a slow write is occurring at the same time the read
comes in.

UniversalSuperBox committed:
Connect saveState as a signal instead of invoking

By registering WindowStateStorage::WindowState as a metatype, we can
skip jumping through the invokeMethod hoop for saveState.

UniversalSuperBox committed:
Test for warnings when AsyncQuery fails to init

peat-psuwit commented on plugins/Utils/windowstatestorage.cpp:
```diff
@@ -16,167 +17,329 @@
+    if (!queryInitSuccessful) {
+        qWarning() << "WindowStateStorage Failed to initialize AsyncQuery! Windows will not be restored to their previous location.";
+        m_asyncOk = false;
+        m_thread.quit();
```
We may want to nullify `m_asyncQuery` here (and then `m_asyncOk` isn't really needed), unless you decided to implement @mariogrip's comment at line 259 (see my comment).

peat-psuwit reviewed (commented):
The thing is, by creating a new `QSqlDatabase` every time, we don't get the benefit of having a dedicated thread, because Qt closes the DB connection as soon as the last `QSqlDatabase` is gone. (Edit: this turns out to be incorrect, see the comment below)

You should be able to just keep `QSqlDatabase` as an `AsyncQuery`'s member, and then replace it with a default-constructed instance before calling `QSqlDatabase::removeDatabase()` in the destructor.

Also...

UniversalSuperBox commented:
From https://doc.qt.io/qt-5/qsqldatabase.html#details (emphasis mine):

> Warning: It is highly recommended that you **do not keep a copy of the QSqlDatabase around as a member of a class**, as this will prevent the instance from being correctly cleaned up on shutdown. **If you need to access an existing QSqlDatabase, it should be accessed with database()**. If you chose to have a QSqlDatabase member variable, this needs to be deleted before the QCoreApplication instance is deleted, otherwise it may lead to undefined behavior.

I don't see any evidence that the connection is closed every time the QSqlDatabase object goes out of scope. Rather, the connection, which is global to the application, is kept open. The `QSqlDatabase` class is only an accessor to QSql's application-global backend.

mardy commented on plugins/Utils/windowstatestorage.cpp:
```diff
@@ -16,167 +17,329 @@
-    const QString queryString = QStringLiteral("INSERT OR REPLACE INTO state (windowId, state) values ('%1', '%2');")
-            .arg(sanitiseString(windowId))
-            .arg((int)state);
+    Q_PROPERTY (const QString dbName READ getDbName)
```
minor: typically properties are declared right below the `Q_OBJECT` macro (whether it's private or public, it does not matter)

mardy commented on plugins/Utils/windowstatestorage.cpp:
```diff
@@ -16,167 +17,329 @@
+    const QString m_getGeometryQuery = QStringLiteral("SELECT * FROM geometry WHERE windowId = :windowId");
+    const QString m_saveGeometryQuery = QStringLiteral("INSERT OR REPLACE INTO geometry (windowId, x, y, width, height) values (:windowId, :x, :y, :width, :height)");
+    const QString m_getStageQuery = QStringLiteral("SELECT stage FROM stage WHERE appId = :appId");
+    const QString m_saveStageQuery = QStringLiteral("INSERT OR REPLACE INTO stage (appId, stage) values (:appId, :stage)");
```
all these can be made `static`

peat-psuwit commented:
I mean, I see this statement from [the destructor's document](https://doc.qt.io/qt-5/qsqldatabase.html#dtor.QSqlDatabase):
> Note: When the last connection is destroyed, the destructor implicitly calls close() to release the database connection.

However, reading the Qt code, apparently there's a `QSqlDatabase` instance in a global cache that will get destroyed when `removeDatabase()` is called. So, this statement, while technically true, is confusing (and misleading). I should file a Qt bug at some point...

UniversalSuperBox committed:
Init AsyncQuery asynchronously

Given review comments to the effect of this being desirable[1],
initialize AsyncQuery using a QueuedConnection instead of a
BlockingQueuedConnection. This makes us move the short-circuit handling
into AsyncQuery.

[1] https://github.com/ubports/unity8/pull/356#discussion_r559277288

UniversalSuperBox committed:
Make AsyncQuery members as static as possible

https://github.com/ubports/unity8/pull/356#discussion_r559752750

UniversalSuperBox committed:
Remove Q_PROPERTY declarations

UniversalSuperBox committed:
Standardize on '&varName' instead of 'TypeName&'

peat-psuwit commented on plugins/Utils/windowstatestorage.h:
```diff
@@ -44,28 +50,25 @@ class WindowStateStorage: public QObject
     Q_FLAG(WindowStates)
 
-    WindowStateStorage(const QString& dbName = nullptr, QObject *parent = nullptr);
+    WindowStateStorage(const QString &dbName = nullptr, QObject *parent = nullptr);
```
One is not supposed to assign or use `nullptr` with a reference. In this case, use an empty `QString()`. This is actually @dobey's fault, but let's fix it here altogether.

```suggestion
    WindowStateStorage(const QString &dbName = QString(), QObject *parent = nullptr);
```

peat-psuwit commented on plugins/Utils/windowstatestorage.cpp:
```diff
@@ -16,167 +17,344 @@
+{
+    qRegisterMetaType<WindowStateStorage::WindowState>("WindowStateStorage::WindowState");
+    QString dbFile;
+    if (dbName == nullptr) {
```
```suggestion
    if (dbName.isEmpty()) {
```

dobey reviewed (approved):
Looks OK.

UniversalSuperBox committed:
Use empty qstring instead of nullptr

https://github.com/ubports/unity8/pull/356#pullrequestreview-572324879

peat-psuwit reviewed (approved):
Should be Ok now.
